core/src/main/java/io/grpc/internal/RetriableStream.java

```diff
@@ -0,0 +1,675 @@
+      substream.writeMessage(messageIs);
+    }
+  }
+
+  private static final class AuthorityEntry implements BufferEntry {
+    final String authority;
+
+    AuthorityEntry(String authority) {
+      this.authority = authority;
+    }
+
+    @Override
+    public void runWith(ClientStream substream) {
+      substream.setAuthority(authority);
+    }
+  }
+
+  private static final class CompressorEntry implements BufferEntry {
+    final Compressor compressor;
+
+    CompressorEntry(Compressor compressor) {
+      this.compressor = compressor;
+    }
+
+    @Override
+    public void runWith(ClientStream substream) {
+      substream.setCompressor(compressor);
+    }
+  }
+
+  private static final class DecompressorRegistryEntry implements BufferEntry {
+    final DecompressorRegistry decompressorRegistry;
+
+    DecompressorRegistryEntry(DecompressorRegistry decompressorRegistry) {
+      this.decompressorRegistry = decompressorRegistry;
+    }
+
+    @Override
+    public void runWith(ClientStream substream) {
+      substream.setDecompressorRegistry(decompressorRegistry);
+    }
+  }
+
+  private static final class FullStreamDecompressionEntry implements BufferEntry {
+    final boolean fullStreamDecompression;
+
+    FullStreamDecompressionEntry(boolean fullStreamDecompression) {
+      this.fullStreamDecompression = fullStreamDecompression;
+    }
+
+    @Override
+    public void runWith(ClientStream substream) {
+      substream.setFullStreamDecompression(fullStreamDecompression);
+    }
+  }
+
+  private static final class MaxInboundMessageSizeEntry implements BufferEntry {
+    final int maxSize;
+
+    MaxInboundMessageSizeEntry(int maxSize) {
+      this.maxSize = maxSize;
+    }
+
+    @Override
+    public void runWith(ClientStream substream) {
+      substream.setMaxInboundMessageSize(maxSize);
+    }
+  }
+
+  private static final class MaxOutboundMessageSizeEntry implements BufferEntry {
+    final int maxSize;
+
+    MaxOutboundMessageSizeEntry(int maxSize) {
+      this.maxSize = maxSize;
+    }
+
+    @Override
+    public void runWith(ClientStream substream) {
+      substream.setMaxOutboundMessageSize(maxSize);
+    }
+  }
+
+  private static final class MessageCompressionEntry implements BufferEntry {
+    final boolean enable;
+
+    MessageCompressionEntry(boolean enable) {
+      this.enable = enable;
+    }
+
+    @Override
+    public void runWith(ClientStream substream) {
+      substream.setMessageCompression(enable);
+    }
+  }
+
+  /** A delayed stream with delayOrExecute() and drain(). */
+  private final class Substream implements ClientStream {
+    final ClientStream stream;
+
+    /** True if all existing buffered callbacks are drained. */
+    volatile boolean upToDate;
+
+    volatile boolean cancelled;
+
+    /** Number of the callbacks in the buffer that have been replayed. */
+    int index = 0;
+
+    List<BufferEntry> buffer = sharedBuffer;
```

Comment:
I didn't think this was going to be necessary. We should maybe discuss why it's needed.